lib/msf/core/payload/malleable_c2.rb

```diff
@@ -0,0 +1,461 @@
+# -*- coding: binary -*-
+
+##
+# This module contains helper functions for parsing and loading malleable
+# C2 profiles into ruby objects.
+##
+
+require 'strscan'
+require 'rex/post/meterpreter/packet'
+
+# Handle escape sequences in the strings provided by the c2 profile
+class String
+  def from_c2_string_value
+    # Support substitution of a subset of escape characters:
+    # \r, \t, \n, \\, \x..
+    # Not supporting \u at this point.
+    # We do in a single regex and parse each as we go, as this avoids the
+    # potential for double-encoding.
+    self.gsub(/\\(x(..)|r|n|t|\\)/) {|b|
+      case b[1]
+      when 'x'
+        [b[2, 4].to_i(16)].pack('C')
+      when 'r'
+        "\r"
+      when 't'
+        "\t"
+      when 'n'
+        "\n"
+      when '\\'
+        "\\"
+      end
+    }
+  end
+end
+
+module Msf::Payload::MalleableC2
+
+  MET = Rex::Post::Meterpreter
+
+  class Token
+    attr_reader :type, :value
+
+    def initialize(type, value)
+      @type = type
+      @value = value
+    end
+  end
+
+  class Lexer
+
+    attr_reader :tokens
+
+    BLOCK_KEYWORDS = %w[
+      client
+      http-get
+      http-post
+      http-stager
+      https-certificate
+      id
+      metadata
+      output
+      server
+      stage
+      transform-x64
+      transform-x86
+    ]
+
+    OTHER_KEYWORDS = %w[
+      add
+      append
+      base64
+      base64url
+      dns
+      encode_hex
+      header
+      hostport
+      mask
+      netbios
+      netbiosu
+      parameter
+      prepend
+      print
+      remove
+      set
+      string
+      stringw
+      strrep
+      transform
+      unset
+      uri
+      uri-append
+      uri-query
+      xor
+    ]
+
+    def initialize(file)
+      @tokens = []
+      tokenize(File.read(file))
```

Comment:
It might not be super needed in this context but we'd preference either:

```suggestion
      tokenize(File.binread(file))
```

Or

```suggestion
      tokenize(File.read(file, mode: 'rb'))
```

To avoid Window's conversion of bytes into new lines which has caused a lot of pain over the years with windows specific file read bugs that contain binary data

> "b"  Binary file mode
>     Suppresses EOL <-> CRLF conversion on Windows. And
>     sets external encoding to ASCII-8BIT unless explicitly
>     specified.
> https://ruby-doc.org/core-2.5.5/IO.html#method-c-read

You can set the encoding yourself if need be

Reply:
`binread` it is!